placement/placement.go

```diff
@@ -126,6 +133,18 @@ func (ps snapshot) MarshalJSON() ([]byte, error) {
 	return json.Marshal(ps.placementSnapshotToJSON())
 }
 
+func (ps *snapshot) UnmarshalJSON(data []byte) error {
```

Comment:
Setting `*ps` in a method seems a bit weird. Perhaps make it a constructor method `NewSnapshotFromJSON(data []byte)`?